### PR TITLE
Add fp16 unit tests

### PR DESCRIPTION
**Pull request by yiqian1**
Add unit tests to ensure that
- database is accessible for tuning parameters of fp16 data
- the lowering pass supports Fp16
- operations of Fp32 also supports Fp16

Fix configuration parameters in two end-to-end test cases to show how to reproduce result failures

### CODE REVIEW

**Comment by whchung** [comment]
Let's see what CI shows.
And I'll abstain from reviewing this PR as I could be biased.

File: mlir/test/Dialect/MIOpen/lowering_filter_tensor_ckyx_cnhw_knhw_f16.mlir [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  return
+}
+// CHECK-LABEL: func @miopen_conv2d
+// CHECK-NEXT:  miopen.transform
+// CHECK:       gridwise_gemm_argument_position = 0
+// CHECK:       names = ["gemmK"]
+// CHECK:       source_names = ["c", "y", "x"]
+// CHECK:       names = ["gemmM"]
+// CHECK:       source_names = ["k"]
+// CHECK-NEXT:  miopen.transform
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/lowering_filter_tensor_ckyx_cnhw_knhw_f16.mlir]: This test is naively added.

Are those logic impacted particularly by data type? If they are not impacted particularly by data type, we should not duplicate the entire `lowering_filter_tensor_ckyx_cnhw_knhw` just for `fp16`. Instead, you should look at the source code and see the aspects that worth adding a fp16 test case and add it to original test case rather than create and duplicate a new one.

Comment (jerryyin) [reply mlir/test/Dialect/MIOpen/lowering_filter_tensor_ckyx_cnhw_knhw_f16.mlir]: Let me make it clear: It looks to me that all test cases in this file is really checking if `memref<*f16>` can be accepted by `-miopen-lowering` stage. If that's the case, what you can do is to create this one test case and only check this additional test case in the `lowering_top_level.mlir` instead of duplicating all the test cases in here.

File: mlir/test/Dialect/MIOpen/lowering_gridwise_gemm_level_f16.mlir [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// This tests checks the following aspects of lowering component:
+// * Can pass arguments correctly 
+// * Can pass arguments in the right sequence
+// * Have the right number of transforms
+// * Have one gridwise_gemm
+
+// RUN: mlir-opt -miopen-lowering %s | FileCheck %s
+
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/lowering_gridwise_gemm_level_f16.mlir]: ditto

File: mlir/test/Dialect/MIOpen/lowering_gridwise_gemm_position_cyxk_chwn_khwn_f16.mlir [hunk 3]
```diff
@@ -0,0 +1,76 @@
+// This tests checks the following aspects of lowering component:
+// * gridwise_gemm argument positions are correct
+
+// RUN: mlir-opt -miopen-lowering %s | FileCheck %s
+
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/lowering_gridwise_gemm_position_cyxk_chwn_khwn_f16.mlir]: ditto

File: mlir/test/Dialect/MIOpen/lowering_input_tensor_cyxk_cnhw_hnhw_f16.mlir [hunk 4]
```diff
@@ -0,0 +1,74 @@
+// This tests checks the following aspects of lowering component:
+// * Input has three transformations in total
+// * Input has correct output_layout across transformations
+
+// RUN: mlir-opt -miopen-lowering %s | FileCheck %s
+
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/lowering_input_tensor_cyxk_cnhw_hnhw_f16.mlir]: ditto

Comment (jerryyin) [reply mlir/test/Dialect/MIOpen/lowering_input_tensor_cyxk_cnhw_hnhw_f16.mlir]: This applies to all `-miopen-lowering` specific test case below.

File: mlir/test/Dialect/MIOpen/lowering_input_tensor_non_zero_padding_f16.mlir [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+// LOWERING-LABEL: func @miopen_conv2d
+// LOWERING:  miopen.transform(%arg1)
+// LOWERING:  output_layout = ["ci", "ni", "hipad", "wipad"]
+// LOWERING:  memref<8x128x34x34xf16>
+
+// AFFINE: #map{{[0-9]+}} = affine_map<(d0, d1) -> (d0 floordiv 9, (d0 mod 9) floordiv 3, (d0 mod 9) mod 3, d1)>
+// AFFINE-NEXT: #map{{[0-9]+}} = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2 - 1, d3 - 1)>
+// AFFINE-NEXT: #map{{[0-9]+}} = affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d2 + d3 - 1, d4 + d5 - 1)>
+// AFFINE-NEXT: #map{{[0-9]+}} = affine_map<(d0, d1) -> (d0 floordiv 9, d1 floordiv 1024, (d0 mod 9) floordiv 3 + (d1 mod 1024) floordiv 32 - 1, (d0 mod 9) mod 3 + (d1 mod 1024) mod 32 - 1)>
+// AFFINE-NEXT: #map{{[0-9]+}} = affine_map<(d0, d1) -> (d0, d1 floordiv 1024, (d1 mod 1024) floordiv 32, (d1 mod 1024) mod 32)>
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/lowering_input_tensor_non_zero_padding_f16.mlir]: None of those affine map change between fp32 and fp16, this test case is not necessary. Lowering aspect has already been covered by other test cases.

**Comment by whchung** [comment]
@yiqian1 The PR has no description. Please supply *WHAT TESTS* are specifically added in this PR.

**Review by whchung — DISMISSED** [review]
The PR lacks description. Please help enumerate what tests are added to reflect fp16-specific tests.

File: mlir/test/Dialect/MIOpen/lowering_top_level.mlir [hunk 6]
```diff
@@ -28,6 +29,28 @@ func @miopen_conv2d(%filter : memref<128x8x3x3xf32>, %input : memref<128x8x32x32
...
+    dilations = [1, 1],
+    strides = [1, 1],
+    padding = [0, 0]
+  } : memref<128x8x3x3xf16>, memref<128x8x32x32xf16>, memref<128x128x30x30xf16>
+  return
+}
+// CHECK-LABEL: func {{@miopen_conv2d.*%arg0.*%arg1.*%arg2}}
+// CHECK-NOT:   miopen.conv2d
+// CHECK-NEXT:  miopen.transform(%arg0)
+// CHECK-NEXT:  miopen.transform(%arg1)
```
Comment (asleepzzz) [inline mlir/test/Dialect/MIOpen/lowering_top_level.mlir:48]: More details for transform?

File: mlir/test/mlir-miopen-driver/e2e/conv2d_host_validation_f16.mlir [hunk 7]
```diff
@@ -63,8 +63,8 @@
...
 
 // COM: bwd_weight
-// FIXME: mlir-miopen-driver -pv -fil_layout=kcyx -in_layout=nchw -out_layout=nkhw  -batchsize=64 -in_channels=4 -out_channels=64 -in_h=32 -in_w=32 -fil_h=1 -fil_w=1   --dilation_h=1 --dilation_w=1 --padding_h=0 --padding_w=0 --conv_stride_h=1 --conv_stride_w=1 -p=false -t f16 --operation=conv2d_bwd_weight -c | mlir-rocm-runner --shared-libs=%rocm_wrapper_library_dir/librocm-runtime-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext --entry-point-result=void | FileCheck %s --check-prefix=CHECK7
+// FIXME: mlir-miopen-driver -pv -fil_layout=kcyx -in_layout=nchw -out_layout=nkhw  -batchsize=256 -in_channels=64 -out_channels=64  -in_h=7 -in_w=7 -fil_h=1 -fil_w=1   --dilation_h=1 --dilation_w=1 --padding_h=0 --padding_w=0 --conv_stride_h=1 --conv_stride_w=1 -p=false -t f16  --operation=conv2d_bwd_weight -c | mlir-rocm-runner --shared-libs=%rocm_wrapper_library_dir/librocm-runtime-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext --entry-point-result=void | FileCheck %s --check-prefix=CHECK7
```
Comment (asleepzzz) [inline mlir/test/mlir-miopen-driver/e2e/conv2d_host_validation_f16.mlir:106]: The parameters changed because of tunning?

Comment (yiqian1) [reply mlir/test/mlir-miopen-driver/e2e/conv2d_host_validation_f16.mlir:106]: The old configuration can not produce valid tuning parameters. I changed to use a set of valid configuration to show the result failure problem.

File: mlir/test/Dialect/MIOpen/Tuning/db_connection.mlir [hunk 8]
```diff
@@ -1,3 +1,5 @@
-// RUN: mlir-miopen-driver -p -miopen-lowering -miopen-affix-params -debug %s 2>&1 | FileCheck %s
+// RUN: mlir-miopen-driver -p -miopen-lowering -miopen-affix-params -debug %s 2>&1 | FileCheck %s --check-prefix=CHECK1
+// RUN: mlir-miopen-driver -p -t f16 -miopen-lowering -miopen-affix-params -debug %s 2>&1 | FileCheck %s --check-prefix=CHECK2
 
-// CHECK: Successfully opened connection to PerfDb
+// CHECK1: Successfully opened connection to PerfDb
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/Tuning/db_connection.mlir]: Please remove all f16 related `Successfully opened connection to PerfDb`. They are payload independent.

You can change some of the existing test case to use f16 but there's no need to add test coverage for it.

File: mlir/test/Dialect/MIOpen/Tuning/db_load_fail.mlir [hunk 9]
```diff
@@ -1,5 +1,9 @@
-// RUN: mlir-miopen-driver -p -miopen-lowering -miopen-affix-params -debug %s 2>&1 | FileCheck %s
+// RUN: mlir-miopen-driver -p -miopen-lowering -miopen-affix-params -debug %s 2>&1 | FileCheck %s --check-prefix=CHECK1
+// RUN: mlir-miopen-driver -p -t f16 -miopen-lowering -miopen-affix-params -debug %s 2>&1 | FileCheck %s --check-prefix=CHECK2
 
-// CHECK: Successfully opened connection to PerfDb
+// CHECK1: Successfully opened connection to PerfDb
 // Perfdb does not have an entry for default configurations
-// CHECK: DB load failed
+// CHECK1: DB load failed
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/Tuning/db_load_fail.mlir]: DB load doesn't consider about the data type as well, only the solverid and performance config. Should remove those redundant one as well.

BTW, can existing solver ids handle fp16 cases?

File: mlir/test/Dialect/MIOpen/Tuning/db_load_success.mlir [hunk 10]
```diff
@@ -1,4 +1,7 @@
-// RUN: mlir-miopen-driver -p=false -batchsize=2 -in_channels=64 -out_channels=256 -in_h=56 -in_w=56 -fil_h=1 -fil_w=1 -dilation_h=1 -dilation_w=1 -conv_stride_h=1 -conv_stride_w=1 -padding_h=0 -padding_w=0 -miopen-lowering -miopen-affix-params -debug %s 2>&1 | FileCheck %s
+// RUN: mlir-miopen-driver -p=false -batchsize=2 -in_channels=64 -out_channels=256 -in_h=56 -in_w=56 -fil_h=1 -fil_w=1 -dilation_h=1 -dilation_w=1 -conv_stride_h=1 -conv_stride_w=1 -padding_h=0 -padding_w=0 -miopen-lowering -miopen-affix-params -debug %s 2>&1 | FileCheck %s --check-prefix=CHECK1
+// RUN: mlir-miopen-driver -p=false -batchsize=2 -in_channels=64 -out_channels=256 -in_h=56 -in_w=56 -fil_h=1 -fil_w=1 -dilation_h=1 -dilation_w=1 -conv_stride_h=1 -conv_stride_w=1 -padding_h=0 -padding_w=0 -t f16 -miopen-lowering -miopen-affix-params -debug %s 2>&1 | FileCheck %s --check-prefix=CHECK2
 
-// CHECK: Successfully opened connection to PerfDb
-// CHECK: DB load succeed
+// CHECK1: Successfully opened connection to PerfDb
+// CHECK1: DB load succeed
+// CHECK2: Successfully opened connection to PerfDb
+// CHECK2: DB load succeed
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/Tuning/db_load_success.mlir]: Same here.

File: mlir/test/Dialect/MIOpen/lowering_top_level.mlir (hunk 6 above)
Comment (jerryyin) [reply mlir/test/Dialect/MIOpen/lowering_top_level.mlir:48]: I think this granularity is good. Is there any specific fp16 related lowering that need to check-out here?

File: mlir/test/Dialect/MIOpen/ops_2_f16.mlir [hunk 11]
```diff
@@ -0,0 +1,232 @@
+// RUN: mlir-opt %s | FileCheck %s
+// RUN: mlir-opt %s | mlir-opt | FileCheck %s
+// Run: mlir-opt -mlir-print-op-generic %s | mlir-opt | FileCheck %s
+
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/ops_2_f16.mlir]: Can we move those tests to two independent test case each with more descriptive name? One for blockwise ops; Another for threadwise ops.

File: mlir/test/Dialect/MIOpen/ops_f16.mlir [hunk 12]
```diff
@@ -0,0 +1,196 @@
+// RUN: mlir-opt %s | FileCheck %s
+// RUN: mlir-opt %s | mlir-opt | FileCheck %s
+// Run: mlir-opt -mlir-print-op-generic %s | mlir-opt | FileCheck %s
+
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/ops_f16.mlir]: There isn't much fp16 dependent logic in those high level ops, consider creating 1 or 2 test case and attach at the corresponding places of the `ops.mlir`

**Review by jerryyin — CHANGES_REQUESTED** [review]

File: mlir/test/Dialect/MIOpen/Tuning/db_load_fail.mlir (hunk 9 above)
Comment (yiqian1) [reply mlir/test/Dialect/MIOpen/Tuning/db_load_fail.mlir]: Do you mean MIOpen solvers? Some of them cannot handle fp16.

**Comment by whchung** [comment]
@yiqian1 ping. Please work with @jerryyin and drive this PR be merged ASAP. I’m not seeing reviewers comments be addressed

File: mlir/test/Dialect/MIOpen/lowering_top_level.mlir (hunk 6 above)
Comment (yiqian1) [reply mlir/test/Dialect/MIOpen/lowering_top_level.mlir:48]: Only threadwiseCopyV2 has one-line code that generated a trunc OP for fp16. But I wasn't able to generate specific test case for it as that Op may get optimized away under some cases.

Comment (jerryyin) [reply mlir/test/Dialect/MIOpen/lowering_top_level.mlir:48]: > as that Op may get optimized away under some cases

At what circumstances does it get optimized away? I do see `FPTruncOp` populated when we do the `threadwiseCopyV2`. However, that is not in this lowering step. This unit test handle to gridwise_gemm lowering only. If this get you confused could you rename those that test this to include `step1` in it. Just so that we know it is `-miopen-lowering-step1`

Comment (yiqian1) [reply mlir/test/Dialect/MIOpen/lowering_top_level.mlir:48]: It happens at step4. It is tested in lowering_threadwise_copy_v2_f16.mlir. The trunc Op gets optimized away when the source and destination data types are the same. I guess the truncation is necessary when the input and output tensors have different date types. But so far we haven't seen that case and I don't know if different data types would be supported. Again I'm not 100% sure when the source and destination types might be different.

Comment (yiqian1) [reply mlir/test/Dialect/MIOpen/lowering_top_level.mlir:48]: Lowering_top_level.mlir just ensures fp16 is handled at the top level.

Comment (jerryyin) [reply mlir/test/Dialect/MIOpen/lowering_top_level.mlir:48]: > It happens at step4. It is tested in lowering_threadwise_copy_v2_f16.mlir.

Make sense. You just confirmed that we don't need additional test case in this unit test.

> But so far we haven't seen that case and I don't know if different data types would be supported. 

https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/021b2dd9c7a22ea563fca19bf281d42390d8eebb/mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h#L4365-L4382

If we are talking about the same thing then it looks to me this truncation always happen for fp16 result type (before the store). Could you add this test case to step4 lowering?

Comment (yiqian1) [reply mlir/test/Dialect/MIOpen/lowering_top_level.mlir:48]: I tried. The trunc Op was generated but didn't show up in the output as it got optimized away.

// *** IR Dump Before MIOpenOpsStep4Pass ***
#map = affine_map<(d0, d1) -> (d0 * 8 + d1)>
module  {
  func @miopen_threadwise_copy_v2(%arg0: i32, %arg1: memref<2xi32, 5>, %arg2: memref<2xi32, 5>, %arg3: vector<32xf16>, %arg4: memref<32xf16>, %arg5: memref<?x?xf16>, %arg6: memref<?x?xf16, #map>, %arg7: memref<?x?x?x?xf16>) {
    %c0 = constant 0 : index
    %c0_i32 = constant 0 : i32
    miopen.threadwise_copy_v2(%arg3, %arg4, %c0_i32, %c0_i32, %c0_i32) {dest_data_per_write = 1 : i64, dim_access_order = [0 : i32], source_data_per_read = 1 : i64, vector_read_write_dim = 0 : i64} : vector<32xf16>, memref<32xf16>
    return
  }
}

 


%5 = vector.extractelement %arg3[%4 : i32] : vector<32xf16>
**%8 = fptrunc %5 : f16 to f16**
store %8, %arg4[%7] : memref<32xf16>
// *** IR Dump After MIOpenOpsStep4Pass ***
#map = affine_map<(d0, d1) -> (d0 * 8 + d1)>
module  {
  func @miopen_threadwise_copy_v2(%arg0: i32, %arg1: memref<2xi32, 5>, %arg2: memref<2xi32, 5>, %arg3: vector<32xf16>, %arg4: memref<32xf16>, %arg5: memref<?x?xf16>, %arg6: memref<?x?xf16, #map>, %arg7: memref<?x?x?x?xf16>) {
    %c0 = constant 0 : index
    %c1 = constant 1 : index
    %c32 = constant 32 : index
    scf.for %arg8 = %c0 to %c32 step %c1 {
      %0 = index_cast %arg8 : index to i32
      %1 = vector.extractelement %arg3[%0 : i32] : vector<32xf16>
      store %1, %arg4[%arg8] : memref<32xf16>
    }
    return
  }
}

Comment (jerryyin) [reply mlir/test/Dialect/MIOpen/lowering_top_level.mlir:48]: Thanks, appreciate helping to check this.

**Review by jerryyin — APPROVED** [review]
